src/target/adiv5_swd.c

```diff
@@ -79,133 +135,200 @@ static uint32_t firmware_dp_low_read(const uint16_t addr)
-			adiv5_dp_write(initial_dp, ADIV5_DP_SELECT, ADIV5_DP_BANK0);
+			adiv5_dp_write(dp, ADIV5_DP_SELECT, ADIV5_DP_BANK2);
+			dp_targetid = adiv5_dp_read(dp, ADIV5_DP_TARGETID);
+			adiv5_dp_write(dp, ADIV5_DP_SELECT, ADIV5_DP_BANK0);
 		}
 	}
 
+	/* If we were given targetid or we have a DPv2+ device, do a multi-drop scan */
+#if PC_HOSTED == 0
+	/* On non hosted platforms, scan_multidrop can be constant */
+	const
+#endif
+		bool scan_multidrop = targetid || dp->version >= 2U;
+
 #if PC_HOSTED == 1
-	if (!initial_dp->dp_low_write) {
-		DEBUG_WARN("CMSIS_DAP < v1.2 can not handle multi-drop, disabling\n");
+	if (scan_multidrop && !dp->dp_low_write) {
+		DEBUG_WARN("Discovered multi-drop enabled target but CMSIS_DAP < v1.2 cannot handle multi-drop\n");
 		scan_multidrop = false;
 	}
 #endif
 
-	DEBUG_WARN("scan_multidrop: %s\n", scan_multidrop ? "true" : "false");
+	if (scan_multidrop)
+		adiv5_swdp_multidrop_scan(dp, dp_targetid);
+	else {
+		adiv5_dp_error(dp);
+		adiv5_dp_init(dp, 0);
+	}
+
+	return target_list ? 1U : 0U;
+}
+
+/*
+ * ARM Debug Interface Architecture Specification, ADIv5.0 to ADIv5.2. ARM IHI 0031C
+ *
+ * §4.2.6 Limitations of multi-drop
+ * 
+ * Each device must be configured with a unique target ID, that includes a 4-bit instance ID, to differentiate between
+ * otherwise identical targets. This places a limit of 16 such targets in any system, and means that identical devices
+ * must be configured before they are connected together to ensure that their instance IDs do not conflict.
+ * Auto-detection of the target
+ * 
+ * It is not possible to interrogate a multi-drop Serial Wire Debug system that includes multiple devices to establish
+ * which devices are connected. Because all devices are selected on coming out of a line reset, no communication with
+ * a device is possible without prior selection of that target using its target ID. Therefore, connection to a multi-drop
+ * Serial Wire Debug system that includes multiple devices requires that either:
+ * - The host has prior knowledge of the devices in the system and is configured before target connection.
+ * - The host attempts auto-detection by issuing a target select command for each of the devices it has been
+ * configured to support.
+ * 
+ * This means that debug tools cannot connect seamlessly to targets in a multi-drop Serial Wire Debug system that they
+ * have never seen before. However, if the debug tools can be provided with the target ID of such targets by the user
+ * then the contents of the target can be auto-detected as normal.
+ * To protect against multiple selected devices all driving the line simultaneously SWD protocol version 2 requires:
+ * - For multi-drop SWJ-DP, the JTAG connection is selected out of powerup reset. JTAG does not drive the line.
+ * - For multi-drop SW-DP, the DP is in the dormant state out of powerup reset.
+ */
+void adiv5_swdp_multidrop_scan(adiv5_debug_port_s *target_dp, const uint32_t targetid)
```

Comment:
`target_dp` here can just be `dp` as we're not sure it adds anything to include the `target_` nomenclature in this instance. Think it can be marked `const` too as in `adiv5_debug_port_s *const dp`

Likewise, please either drop the `p` from the end of `swdp` or make it `swd_dp` - we've been working to slowly eliminate this confusing naming, so not adding another instance to later fix is a good idea.

Reply:
naming suggestions taken, regarding the const, it cant as that same pointer was being used to track the new dp allocations, but if I use the dp allocation strategy suggested in the other comment I might be able to